video/out/opengl/shader_cache.c

```diff
@@ -239,6 +258,14 @@ static int gl_sc_next_binding(struct gl_shader_cache *sc, enum ra_vartype type)
     }
 }
 
+static int gl_sc_next_offset(struct gl_shader_cache *sc, size_t el_size,
+                             int num)
+{
+    size_t offset = MP_ALIGN_UP(sc->ubo_size, el_size);
+    sc->ubo_size = offset + el_size * num;
+    return offset;
```

Comment:
Are you sure that's sufficient for std140?

Reply:
Not sure. The rules here seem to agree:

![std140](https://www.safaribooksonline.com/library/view/opengl-programming-guide/9780132748445/graphics/app09tab01.jpg)

Reply:
The way I read that is that `float[2]` is as large as 2*vec4, and that can't be. (It says each array element's size is rounded up to a multiple of the size of vec4.)

Reply:
> The way I read that is that float[2] is as large as 2*vec4, and that can't be.

That's correct, though. That's how std140 works, and why std430 was invented. (The only change between std430 and std140 is that arrays are "densely packed" in std430)

Reply:
That's dumb. I don't think your code does that? Also ubo_size might lack padding for the last element.

Reply:
> I don't think your code does that?

```
    if (u->input.dim_m > 1)
        u->ubo_stride = MP_ALIGN_UP(u->ubo_stride, sizeof(float[4]));
```

It does so here.

> Also ubo_size might lack padding for the last element.

`sc->ubo_size = u->ubo_offset + u->ubo_stride * u->ubo_length;`
ubo_stride includes all the necessary padding for each element